src/server/service/roles/db.js

```diff
@@ -0,0 +1,78 @@
+const { Pool } = require('pg');
+
+const { sql } = require('../../util/sqlHelpers');
+
+const pool = new Pool();
+
+exports.getUserRoles = async function getUserRoles(userId) {
+    const client = await pool.connect();
+    try {
+        const result = await client.query(
+            sql`SELECT * FROM user_role WHERE user_id = ${userId};`
+        );
+        return { roles: result.rows.map(row => row.role) };
+    } finally {
+        client.release();
+    }
+};
+
+exports.addUserRoles = async function addUserRoles(userId, roles) {
+    const client = await pool.connect();
+    try {
+        await client.query('BEGIN');
+        const result = await client.query(sql`
+INSERT INTO user_role (user_id, role)
+    SELECT ${userId} as user_id, t.role FROM jsonb_array_elements_text(${roles}) AS t (role)
```

Comment:
I'm not sure I fully follow what this line is doing. Are we extracting each role in the array and adding a row for it? Or all of the roles are an array in one row?

Reply:
pop this into psql -- this is effectively what the SELECT statement does:
```
teachermoments=> SELECT
teachermoments->   1 as user_id,
teachermoments->   t.role
teachermoments-> FROM jsonb_array_elements_text('["admin", "super_admin"]'::jsonb) AS t (role);
 user_id |    role
---------+-------------
       1 | admin
       1 | super_admin
(2 rows)
```

jsonb_array_elements_text(jsonbarray) returns a "setof text" which is basically a "select result" or a "table" -- so you can alias it using `AS t` and then define the column name to make the `t.role` field that was a part of the SELECT  earlier, making it  `as t (role)`

so each element in the array becomes a value that we try to insert into the database paired with the users id!

Reply:
Neat!